src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java

```diff
@@ -61,24 +60,35 @@ public Stream<BitbucketTag> getRemoteTags() {
                 .flatMap(Collection::stream);
     }
 
-    public BitbucketTag getRemoteTag(String tagName) {
-        HttpUrl.Builder urlBuilder = null;
-        try {
-            urlBuilder = bitbucketRequestExecutor.getCoreRestPath().newBuilder()
+    @Override
+    public BitbucketTag getRemoteTag(String tagName) throws NotFoundException {
+        HttpUrl url = bitbucketRequestExecutor.getCoreRestPath().newBuilder()
                     .addPathSegment("projects")
                     .addPathSegment(projectKey)
                     .addPathSegment("repos")
                     .addPathSegment(repositorySlug)
-                    .addPathSegment("tags")
-                    .addPathSegment(URLEncoder.encode(tagName, "UTF-8"));
-        } catch (UnsupportedEncodingException e) {
-            throw new RuntimeException(e);
+                    .addPathSegment("commits")
+                    .addQueryParameter("until", tagName)
+                    .addQueryParameter("start", "0")
+                    .addQueryParameter("limit", "1")
+                    .build();
+
+        BitbucketPage<BitbucketCommit> firstPage =
+                bitbucketRequestExecutor.makeGetRequest(url,
+                        new TypeReference<BitbucketPage<BitbucketCommit>>() {
+                        }).getBody();
+
+        BitbucketCommit commit = BitbucketPageStreamUtil.toStream(firstPage, new BitbucketTagClientImpl.OnlyPageFetcherImpl())
+                .map(BitbucketPage::getValues)
+                .flatMap(Collection::stream)
+                .findFirst().orElse(null);
+
+        if (commit == null) {
+            throw new NotFoundException(String.format("Unable to locate tag with name %s", tagName), firstPage.toString());
         }
 
-        HttpUrl url = urlBuilder.build();
-        BitbucketTag tag = bitbucketRequestExecutor.makeGetRequest(url, BitbucketTag.class).getBody();
+        return new BitbucketTag(commit.getId(), commit.getDisplayId(), commit.getId());
```

Comment:
This feels a bit odd to me. We don't really check the tag name, so if a commit hash was passed to this method it'd work fine and return as though the hash was a tag.
Not a huge deal because this is internal, but it looks like this maps the commit and just returns the `latestCommit`, which here is just `commit.getId`. So I wonder if this method should just be `getLatestCommitFromTag`, and you can return a `BitbucketCommit` exactly as it's unmarshalled? It actually saves you a step in the SCM source and feels a bit more honest in what we're doing here.

Reply:
We can just slightly modify the getLatestCommit code to take a `BitbucketScmHead` and then reuse the function for tags. I think thats the cleaner way forward

Reply:
Works for me!